### ISSUE TITLE
[SP8192 + 3-Layer Recurrence + Parallel Residuals + Legal TTT] Seed 42 log contains ttt_hash_embed/ttt_hash_buckets not present in submitted train_gpt.py

### ISSUE THREAD

**Issue by GabrieleCirillo** [openai/parameter-golf#1594]
## `ttt_hash_embed` and `ttt_hash_buckets` present in seed 42 log but not in other seeds or `train_gpt.py`

**Record:** `records/track_10min_16mb/2026-04-09_SP8192_3LayerRecur_ParResid_QK525_LegalTTT`

While running experiments based on this record, I noticed a discrepancy across the three seed logs. Seed 42 includes two hyperparameters — `ttt_hash_embed: True` and `ttt_hash_buckets: 16384` — that don't appear in seeds 314 or 999, and don't appear anywhere in the committed `train_gpt.py`:

| Field | Seed 42 | Seed 314 | Seed 999 |
|---|---|---|---|
| `ttt_hash_embed` | `True` | not present | not present |
| `ttt_hash_buckets` | `16384` | not present | not present |
| `Code size` | 16630 bytes | 16594 bytes | 16594 bytes |
| `quantized_ttt val_bpb` | 1.08079 | 1.08103 | 1.08118 |

Seeds 314 and 999 both report `Code size: 16594 bytes`, consistent with the submitted code. Seed 42 reports 16630 bytes, suggesting it was run with a different version of `train_gpt.py` that includes the `ttt_hash_embed` / `ttt_hash_buckets` feature.

A question:
Could the version of `train_gpt.py` used for seed 42 be shared? I'm also curious about `ttt_hash_embed` and `ttt_hash_buckets` as techniques.

**Comment by himanshudongre** [comment]
Thanks for flagging this, @GabrieleCirillo — I spent today reproducing this record from the committed code and ran into exactly the gap you describe, so wanted to share three additional pieces of evidence that build on your observation. They point fairly strongly toward a specific earlier-merged implementation as the most likely source of `ttt_hash_embed` / `ttt_hash_buckets`.

**TL;DR**: the feature looks like it matches `BigramHashEmbedding` from the earlier merged record `records/track_10min_16mb/2026-03-20_Int6_MLP3x_SmearGate_BigramHash_MuonWD_SWA`. Reproducing from the committed `train_gpt.py` on 8×H100 gets **1.143 BPB** vs the log's 1.081 — a 0.062-BPB gap whose growth pattern (identical at init, widening during training) is consistent with a zero-initialized auxiliary embedding being silently active in the seed-42 run.

---

## Evidence

### 1. Committed code contains zero implementation of the hashed feature

`train_gpt.py` submitted with PR #1493 (decoded from its `B.b85decode(...)` + `lzma.decompress(..., FORMAT_RAW, LZMA2)` blob — 48,583 bytes, 469 lines):

```
$ grep -nE 'hash|bucket|bigram' /tmp/pr1493_decoded.py
(no output)
```

No class, no function, no string reference — the committed training script has no code path that uses `ttt_hash_embed` or `ttt_hash_buckets` regardless of their hparam values.

### 2. The winning seed-42 log references both hparams

From `records/track_10min_16mb/2026-04-09_SP8192_3LayerRecur_ParResid_QK525_LegalTTT/train_seed42.log`:

```
  ttt_chunk_tokens: 32768
  ttt_enabled: True
  ttt_epochs: 3
  ttt_hash_buckets: 16384     <-- referenced here
  ttt_hash_embed: True        <-- referenced here
  ttt_lr: 0.005
  ttt_momentum: 0.9
```

These print only because the training script that actually ran had those attributes on its `Hyperparameters` dataclass. The committed script does not.

### 3. A previously-merged record contains the matching implementation

`records/track_10min_16mb/2026-03-20_Int6_MLP3x_SmearGate_BigramHash_MuonWD_SWA/train_gpt.py` (lines 584-608) defines:

```python
class BigramHashEmbedding(nn.Module):
    """Hash consecutive token pairs into a learned embedding table."""
    def __init__(self, bigram_vocab_size, bigram_dim, model_dim):
        super().__init__()
        self.bigram_vocab_size = bigram_vocab_size
        self.embed = nn.Embedding(bigram_vocab_size, bigram_dim)
        nn.init.zeros_(self.embed.weight)
        self.proj = CastedLinear(bigram_dim, model_dim, bias=False) if bigram_dim != model_dim else None
        if self.proj is not None:
            nn.init.zeros_(self.proj.weight)
        self.scale = nn.Parameter(torch.tensor(0.05, dtype=torch.float32))

    def bigram_hash(self, tokens):
        t = tokens.to(torch.int32); mod = self.bigram_vocab_size - 1
        out = torch.empty_like(t)
        out[..., 0] = mod
        out[..., 1:] = torch.bitwise_xor(36313 * t[..., 1:], 27191 * t[..., :-1]) % mod
        return out.long()

    def forward(self, token_ids):
        h = self.embed(self.bigram_hash(token_ids))
        if self.proj is not None: h = self.proj(h)
        return h * self.scale.to(dtype=h.dtype)
```

Three coincidences all line up:
- `bigram_vocab_size` in that class = `ttt_hash_buckets` in PR #1493's log (both 16384 capable).
- The class is *zero-initialized* (embed weights, proj weights), so a model with it active but freshly initialized produces the **same step-1 loss** as a model without it. This explains why a blind reproduction from the committed PR #1493 code tracks their log exactly at step 1 (9.0104 vs 9.0111) but diverges by step 4000.
- The filesystem-adjacent merged record is named `BigramHash_MuonWD_SWA` — the author's own prior work.

### 4. Stage-by-stage reproduction gap (8×H100, SEED=42, committed code only)

| Stage                         | PR #1493 log | My reproduction  | Gap     |
| ----------------------------- | -----------: | ---------------: | ------: |
| step 1 train_loss             | 9.0111       | 9.0104           | ≈0.0007 |
| step 4000 val_bpb             | 1.1119       | 1.1514 @ 4317    | +0.040  |
| final (wallclock_cap) val_bpb | 1.0886       | 1.1514           | +0.063  |
| pre-quantization post-EMA     | 1.0873       | 1.1502           | +0.063  |
| quantized                     | 1.0997       | 1.1620           | +0.062  |
| quantized + sliding window    | 1.0829       | 1.1444           | +0.062  |
| quantized + TTT               | 1.0808       | 1.1431           | +0.062  |

Identical at initialization (same seed, same architecture), divergence grows as training proceeds — the expected signature of a silently-active zero-initialized auxiliary embedding that the reproduced model lacks.

(Model parameter count printed by both runs: `35,944,536` — identical. Either both counts exclude the bigram table, or the committed script's `model_params` log line is computed before the silently-added layer. Either way, the architectural mismatch is invisible to a param-count check.)

---

## Why this matters for reproducers

Anyone trying to **build on** this record by starting from the committed training script will hit the same ~1.143 ceiling I did — the committed script has no code path that can use `ttt_hash_embed` regardless of what value is set, and the gap isn't visible to a parameter-count check (both runs print `model_params: 35,944,536`).

Flagging this specifically so other folks don't unnecessarily burn cloud-GPU time the way I did. What made it expensive to catch is the zero-init property — the step-1 train_loss matches PR #1493's log to 3 decimals (9.0104 vs 9.0111), so a reproduction *looks* like it's tracking the reference correctly at the start. The natural reaction is to assume your own code or environment is at fault and keep re-running additional seeds, which is what I did before finally grep'ing the committed script. If the full training code were committed, future reproducers wouldn't hit the same trap.

## Questions for the maintainers / PR author

Tagging @bigbag (PR #1493 author) since these mostly need your knowledge — I'd genuinely appreciate clarification on any of the following, and I'm happy to contribute a PR myself if useful:

1. **Is `BigramHashEmbedding` the actual feature behind `ttt_hash_embed`?** Given the name, the 16384 bucket size match, and the zero-init behavior, it seems the most likely candidate — but you'd know for sure.
2. **Would it be useful to commit an updated `train_gpt.py` for this record** that includes the class (or equivalent), so the seed-42 log is reproducible end-to-end from source? Glad to open that PR if you prefer — happy to use whatever hparam names you'd like (`bigram_vocab_size` / `bigram_dim`, or keeping `ttt_hash_*`).
3. **For future records** — @cocohearts, is there a maintainer policy preference on whether silently-inherited features from prior merged records should be called out, vs. always committing a fully self-contained script? Even a one-line note in the README about reproducibility expectations would save future submitters (and reviewers) from re-running into this.

Again, thanks @GabrieleCirillo for surfacing this, and thanks @bigbag for the record itself — the recipe is clearly strong, just trying to help close the loop so downstream work can build on it more easily.

---

**Artifacts referenced**:
- `/tmp/pr1493_decoded.py` — decoded from `records/track_10min_16mb/2026-04-09_SP8192_3LayerRecur_ParResid_QK525_LegalTTT/train_gpt.py` via `lzma.decompress(base64.b85decode(...), format=lzma.FORMAT_RAW, filters=[{'id': lzma.FILTER_LZMA2}])`
- `records/track_10min_16mb/2026-04-09_SP8192_3LayerRecur_ParResid_QK525_LegalTTT/train_seed42.log` (unchanged)
- `records/track_10min_16mb/2026-03-20_Int6_MLP3x_SmearGate_BigramHash_MuonWD_SWA/train_gpt.py` lines 584-608, 645-671, 688-690, 714-716, 914-944

**Comment by codemath3000** [comment]
Quick note: this issue does **not** affect my performance-optimized PR (#1583) based on this submission. My PR runs the same code for every seed, as can be verified in my logs, and is fully reproducible.
